dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java

```diff
@@ -340,101 +437,76 @@ public ImportSummaries addEvents( List<Event> events, ImportOptions importOption
     }
 
     @Override
-    public ImportSummary addEvent( Event event, ImportOptions importOptions )
+    public ImportSummary addEvent( Event event, ImportOptions importOptions, boolean bulkImport )
     {
         importOptions = updateImportOptions( importOptions );
+        
+        ProgramStageInstance programStageInstance = getProgramStageInstance( event.getEvent() );
 
-        if ( programStageInstanceService.programStageInstanceExistsIncludingDeleted( event.getEvent() ) )
+        if (  programStageInstance != null && programStageInstance.isDeleted() )
         {
-            return new ImportSummary( ImportStatus.ERROR, "Event ID " + event.getEvent() + " was already used. The ID is unique and cannot be used more than once" )
+            return new ImportSummary( ImportStatus.ERROR, "Event ID " + event.getEvent() + " was already used and deleted. This event can not be modified." )
                 .setReference( event.getEvent() ).incrementIgnored();
         }
+        
+        if ( programStageInstance == null && !StringUtils.isEmpty( event.getEvent() ) && !CodeGenerator.isValidUid( event.getEvent() ) )
```

Comment:
Wondering about this IF statement:
- PSI was not found in DB - OK
AND
- UID in the payload is present
AND
- UID in the payload is not in the correct state

Why does it have to be in the invalid state?

If PSI is found, we skip IF branch and continue. But then it is not addEvent anymore but updateEvent. Or?

Basically, I would expect this logic:
1. If it is present in DB and deleted -> error
2. If it is present in DB and not deleted -> error (because this is addEvent and not updateEvent method)
3. If it is not present in DB -> continue
3.1 If Event UID is present in the payload, validate whether it is OK: NO -> Fail, Yes -> use it
3.2 If UID is not present in the payload -> fine, go on, I will generate a new one at the time when I need it and all validation is passed.

Reply:
rejecting invalid UID is what we had from before. I guess the intention is - that invalid UID might have been referenced somewhere, scanning and replacing with a valid UID would be a difficult task. May be @mortenoh can comment on this... what are we doing with invalid UIDs during metadata import?